### ISSUE TITLE
linter: avoid `prefer-optional-chain` breaking symmetry

### ISSUE THREAD

**Issue by jeysal** [oxc-project/oxc#21849]
### What version of Oxlint are you using?

1.61.0

### What command did you run?

oxc neovim lsp

### What does your `.oxlintrc.json` (or `oxlint.config.ts`) config file look like?

complicated generated config at @framer

### What happened?

More of a rule improvement than a bug to prevent certain false positives.

```
function test() {
	let x: undefined | { a: string }, y: undefined | { a: string }
	if (x && y && x.a === y.a) {
		console.log("x and y are equal")
	}
}
```

`prefer-optional-chain` flags this `if`, and it does correctly stop flagging when converting one of `x` or `y` to optional chaining (to prevent `undefined === undefined`).
But it's not ideal to flag at all because we don't want to break the symmetry here for human readability.
One way that this case can be prevented/detected (even without extra type info checks) if I'm not mistaken is e.g. by recursively applying the rule:

Apply rule to whole condition, do not stop after detecting one flag. IF rule flags `x` *and* rule flags `y` AND in the modified tree after autofixing `x` it doesn't flag `y` anymore AND in the modified tree after autofixing `y` it doesn't flag `x` anymore. (mutual exclusivity)
In that case, we don't want either flagged in the first place.

Idk if y'all have the infra to implement that easily, but wanted to suggest the improvement in any case. There might also be an easier way to detect it than my spontaneously conceived algo.

**Comment by camc314** [comment]
Thanks for the report - oxlint-tsgolint v0.22.1 is releasing now
